### PR TITLE
[Model] Add support for H2OVL-Mississippi models

### PR DESCRIPTION
**Pull request by cooleel**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

- This PR adds support for H2OVL-Mississippi series models:
Collection: https://huggingface.co/collections/h2oai/h2ovl-mississippi-66e492da45da0a1b7ea7cf39
h2oai/h2ovl-mississippi-800m: https://huggingface.co/h2oai/h2ovl-mississippi-800m
h2oai/h2ovl-mississippi-2b: https://huggingface.co/h2oai/h2ovl-mississippi-2b

- The closest model vllm already supports.
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/internvl.py

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/h2ovl.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+            ((i // (target_width // image_size)) + 1) * image_size,
+        )
+        # split the image
+        split_img = resized_img.crop(box)
+        processed_images.append(split_img)
+    assert len(processed_images) == blocks
+    if use_thumbnail and len(processed_images) != 1:
+        thumbnail_img = image.resize((image_size, image_size))
+        processed_images.append(thumbnail_img)
+    return processed_images, target_aspect_ratio
```
Comment (Isotr0py) [inline vllm/model_executor/models/h2ovl.py:106]: Since `dynamic_preprocess2` has exposed and processed `prior_aspect_ratio` conditionally, is this still necessary?

File: vllm/model_executor/models/h2ovl.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+
+# new dynamic_preprocess2 with prior_aspect_ratio
+def dynamic_preprocess2(
+    image: Image.Image,
+    min_num: int,
+    max_num: int,
+    image_size: int,
+    use_thumbnail: bool,
+    prior_aspect_ratio: Tuple[int, int],
```
Comment (Isotr0py) [inline vllm/model_executor/models/h2ovl.py:116]: I think we can set `prior_aspect_ratio=None` here so that it can work just like `dynamic_preprocess` above.

**Review by Isotr0py — COMMENTED** [review]
Some initial comments about twice image processing. Seems that `load_image1`/`load_image2` and `dynamic_preprocess1`/`dynamic_preprocess2` are a little bit duplicated.

File: vllm/model_executor/models/h2ovl.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+        transform = build_transform(input_size=input_size)
+        images, _ = dynamic_preprocess(
+            image,
+            min_num=min_num,
+            max_num=max_num,
+            image_size=input_size,
+            use_thumbnail=use_thumbnail,
+        )
+        pixel_values = [transform(image) for image in images]
+        pixel_values = torch.stack(pixel_values)
```
Comment (Isotr0py) [inline vllm/model_executor/models/h2ovl.py:221]: This seems like the duplicate implementation of `load_image1`. Why can't we just call `load_image1(...)` here?

File: vllm/model_executor/models/h2ovl.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+        image,
+        image_size=input_size,
+        use_thumbnail=True,
+        min_num=min_num,
+        max_num=max_num,
+        prior_aspect_ratio=target_aspect_ratio,
+    )
+    pixel_values = [transform(image) for image in images]
+    pixel_values = torch.stack(pixel_values)
+    return pixel_values
```
Comment (Isotr0py) [inline vllm/model_executor/models/h2ovl.py:184]: Seems that `load_image2` works just like `load_image1` if `target_aspect_ratio=None`, only differs from retrun values. Can we replace `load_image1` with `load_image2` with the modification of `dynamic_preprocess`?

File: vllm/model_executor/models/h2ovl.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+        model_config = ctx.model_config
+        hf_config = ctx.get_hf_config()
+
+        image_data = multi_modal_data["image"]
+        num_patches = get_internvl_num_patches(hf_config)
+
+        # can only get the total blocks num after the image fully processed
+        num_blocks_calculator = image_to_pixel_values_wrapper(
+            hf_config, max_dynamic_patch=max_dynamic_patch)
```
Comment (ywang96) [inline vllm/model_executor/models/h2ovl.py:289]: We should definitely avoid calling the image processing twice. If there's no way to get around this, you could also return the processed pixel values within `input_processor`, and make `input_mapper` a no-op function. 

Molmo has done something similar here https://github.com/vllm-project/vllm/blob/eae3d48181b1ad27f132f14df18e8cff203f7552/vllm/model_executor/models/molmo.py#L860-L866

File: vllm/model_executor/models/h2ovl.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+        image_pixel_values_mapper = image_to_pixel_values_wrapper(
+            hf_config, max_dynamic_patch)
+
+        if isinstance(data, Image.Image):
+            data = image_pixel_values_mapper(data)
+            data = data.unsqueeze(0)
+        elif is_list_of(data, Image.Image):
+            hf_config.use_msac = False
+            data = [image_pixel_values_mapper(img) for img in data]
```
Comment (ywang96) [inline vllm/model_executor/models/h2ovl.py:346]: Per my comment above - this second call of image processing is what we should avoid

**Review by ywang96 — CHANGES_REQUESTED** [review]
Hello @cooleel! Thank you for making this contribution!

Please take a look at the comments, especially those regarding how we should avoid double image processing calls.

File: vllm/model_executor/models/h2ovl.py (hunk 1 above)
Comment (cooleel) [reply vllm/model_executor/models/h2ovl.py]: The duplicated implementation is fixed.

File: vllm/model_executor/models/h2ovl.py (hunk 4 above)
Comment (cooleel) [reply vllm/model_executor/models/h2ovl.py]: The duplicated implementation is fixed.

File: vllm/model_executor/models/h2ovl.py (hunk 5 above)
Comment (cooleel) [reply vllm/model_executor/models/h2ovl.py]: Thanks for the suggestion, I followed the approach in molmo.py to prepare the `multi_modal_data` within `input_processor`, so there’s no processing for real image data in input_mapper. I’ve kept the image processing code for dummy_data, as I intend to reuse the dummy_data from InternVL.

**Comment by cooleel** [comment]
Thanks for the careful review and insightful comments.
I've made several improvements and refactored the code to address the points you raised. 
Please take a look, and let me know if further adjustments are needed.

File: vllm/model_executor/models/h2ovl.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+    return num_blocks * num_patches
+
+
+class H2OVLInputPipeline(InternVLInputPipeline):
+    """
+    Input pipeline for processing image and text data for the H2OVL model.
+    """
+
+    def __init__(self):
+        super().__init__(IMG_START, IMG_END, IMG_CONTEXT)
```
Comment (Isotr0py) [inline vllm/model_executor/models/h2ovl.py:222]: I think this is unnecessary, you can just call `H2OVLInputPipeline(IMG_START, IMG_END, IMG_CONTEXT)`

File: vllm/model_executor/models/h2ovl.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+            pixel_values = pixel_values.unsqueeze(0)
+
+        # multi images
+        elif is_list_of(image_data, Image.Image):
+            # Do not use MSAC for multi images
+            hf_config.use_msac = False
+            image_feature_sizes = []
+            pixel_values = [
+                image_pixel_values_mapper(image) for image in image_data
+            ]
```
Comment (Isotr0py) [inline vllm/model_executor/models/h2ovl.py:259]: If MSAC is not used for multi images, you should expose the `use_MSAC` for `image_pixel_values_mapper` instead of override the hf_config:
```python
image_pixel_values_mapper(image, use_MSAC=False)
```

File: vllm/model_executor/models/h2ovl.py (hunk 7 above)
Comment (cooleel) [reply vllm/model_executor/models/h2ovl.py]: Fixed.

File: vllm/model_executor/models/h2ovl.py (hunk 8 above)
Comment (cooleel) [reply vllm/model_executor/models/h2ovl.py]: Fixed.

**Comment by cooleel** [comment]
Code was formatted by format.sh. but still failed in yapf.  mentioned in #9890

**Comment by ywang96** [comment]
> Code was formatted by format.sh. but still failed in yapf. mentioned in #9890

This is an interesting one - yapf and isort are actually overwriting each other..

**Comment by ywang96** [comment]
> @ywang96 Thank you for working through the format tests! If you spot any areas in my code that need further adjustment, please feel free to let me know.

Thanks! I will give this PR some testing but we should be able to merge this by EoD!

**Comment by cooleel** [comment]
Hi, @ywang96  I've updated this branch with the latest changes from the main branch.  Let me know if you encounter any issues in tests or if there's anything else I can assist with. Thank you for your time and help!

**Review by ywang96 — APPROVED** [review]
@cooleel Thanks for your contribution! I updated this PR so that it should work properly now with image embeddings as input. Otherwise everything looks good to me!
